mongo-utils/src/main/java/org/wso2/mongo/MongoDBDataSourceUtil.java

```diff
@@ -202,77 +221,100 @@ private static AuthenticationMechanism retrieveAuthMechanism(String authMechanis
 
     private static MongoClientOptions createOptions(BMap options) {
         MongoClientOptions.Builder builder = MongoClientOptions.builder();
-        boolean sslEnabled = options.getBooleanValue(ConnectionParam.SSL_ENABLED.getKey());
-        if (sslEnabled) {
-            builder = builder.sslEnabled(true);
-            boolean sslInvalidHostNameAllowed = options.getBooleanValue(ConnectionParam.SSL_INVALID_HOSTNAME_ALLOWED
-                    .getKey());
-            if (sslInvalidHostNameAllowed) {
-                builder.sslInvalidHostNameAllowed(true);
+        if (options != null) {
+            boolean sslEnabled = options.containsKey(ConnectionParam.SSL_ENABLED.getKey()) ? 
+                                options.getBooleanValue(ConnectionParam.SSL_ENABLED.getKey()) : false;
+            if (sslEnabled) {
+                builder = builder.sslEnabled(true);
+                boolean sslInvalidHostNameAllowed = options.containsKey(ConnectionParam.SSL_INVALID_HOSTNAME_ALLOWED
+                                .getKey()) ? options.getBooleanValue(ConnectionParam.SSL_INVALID_HOSTNAME_ALLOWED
+                                .getKey()) : false;
+                if (sslInvalidHostNameAllowed) {
+                    builder.sslInvalidHostNameAllowed(true);
+                }
+                builder.sslContext(initializeSSLContext(options));
+            }
+            if (options.containsKey(ConnectionParam.RETRY_WRITES.getKey())) {
+                builder.retryWrites(options.getBooleanValue(ConnectionParam.RETRY_WRITES.getKey()));
+            }
+            if (options.containsKey(ConnectionParam.READ_CONCERN.getKey())) {
+                String readConcern = options.getStringValue(ConnectionParam.READ_CONCERN.getKey()).getValue();
+                if (!readConcern.isEmpty()) {
+                    builder = builder.readConcern(new ReadConcern(ReadConcernLevel.valueOf(readConcern)));
+                }            
+            }
+            if (options.containsKey(ConnectionParam.WRITE_CONCERN.getKey())) {
+                String writeConcern = options.getStringValue(ConnectionParam.WRITE_CONCERN.getKey()).getValue();
+                if (!writeConcern.isEmpty()) {
+                    builder = builder.writeConcern(WriteConcern.valueOf(writeConcern));
+                }
+            }
+            
+            if (options.containsKey(ConnectionParam.READ_PREFERENCE.getKey())) {
+                String readPreference = options.getStringValue(ConnectionParam.READ_PREFERENCE.getKey()).getValue();
```

Comment:
We can use a single util method to do this validation. 

privete string getValueFromBMap(BMap map, string key) {..}

Reply:
If we use util method , it will be bit complicated. Again we need to use this lengthy line of code t call the methods. As we discussed offline, let's keep this as it is.